x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/action/PostAnalyticsEventAction.java

```diff
@@ -0,0 +1,248 @@
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.application.analytics.action;
+
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.action.ActionType;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.logging.LoggerMessageFormat;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentType;
+import org.elasticsearch.xpack.application.analytics.event.AnalyticsEvent;
+import org.elasticsearch.xpack.application.analytics.event.AnalyticsEventFactory;
+
+import java.io.IOException;
+import java.util.Locale;
+import java.util.Objects;
+
+import static org.elasticsearch.action.ValidateActions.addValidationError;
+
+public class PostAnalyticsEventAction extends ActionType<PostAnalyticsEventAction.Response> {
+
+    public static final PostAnalyticsEventAction INSTANCE = new PostAnalyticsEventAction();
+
+    public static final String NAME = "cluster:admin/xpack/application/analytics/post_event";
+
+    private PostAnalyticsEventAction() {
+        super(NAME, Response::new);
+    }
+
+    public static class Request extends ActionRequest implements AnalyticsEvent.Context {
+
+        private final String eventCollectionName;
+
+        private final String eventType;
+
+        private final long eventTime;
+
+        private final boolean debug;
+
+        private final BytesReference payload;
+
+        private final XContentType xContentType;
+
+        public Request(String eventCollectionName, String eventType, boolean debug, XContentType xContentType, BytesReference payload) {
+            this(eventCollectionName, eventType, debug, System.currentTimeMillis(), xContentType, payload);
+        }
+
+        public Request(
+            String eventCollectionName,
+            String eventType,
+            boolean debug,
+            long eventTime,
+            XContentType xContentType,
+            BytesReference payload
+        ) {
+            this.eventCollectionName = eventCollectionName;
+            this.eventType = eventType;
+            this.debug = debug;
+            this.eventTime = eventTime;
+            this.xContentType = xContentType;
+            this.payload = payload;
+        }
+
+        public Request(StreamInput in) throws IOException {
+            super(in);
+            this.eventCollectionName = in.readString();
+            this.eventType = in.readString();
+            this.debug = in.readBoolean();
+            this.eventTime = in.readLong();
```

Comment:
Are we including a timestamp for when the event has been received as well, which could be auto-generated by the server?

Reply:
The constructor that is called by the rest handler is:

```
public Request(String eventCollectionName, String eventType, boolean debug, XContentType xContentType, BytesReference payload) {
   this(eventCollectionName, eventType, debug, System.currentTimeMillis(), xContentType, payload);
}
```

It generate the eventTime as soon as the request is received and use `System.currentTimeMillis()`.

The additional constructor are meant to be use for tests / deserialize request in the transport layer.

Reply:
This makes me thinks that we should expect the eventTime to be set client side and be part of the sent event?